File: generate/reference/duality-of-compilation/ocaml/iterate_increment.ml
let rec iterate i f a =
  if i = 0 then
    a
  else
    iterate (i - 1) f (f a)

let main n =
  iterate n (fun x -> x + 1) 0

let () =
  match Sys.argv with
  | [| _ |] -> print_endline (string_of_int (main 10))
  | [| _; s |] -> (
      match int_of_string_opt s with
      | Some n -> print_endline (string_of_int (main n))
      | None -> print_endline "argument is not a number")
  | _ -> print_endline "arguments are too many"

File: generate/reference/duality-of-compilation/rust/iterate_increment.rs
use std::env;

fn iterate(i: isize, f: impl Fn(isize) -> isize, arg: isize) -> isize {
    let mut res = arg;
    for _ in 0..i {
        res = f(res);
    }
    res
}

fn run(n: isize) -> isize {
    iterate(n, |x| x + 1, 0)
}

fn main() {
    let args: Vec<String> = env::args().collect();

    let arg = match args.get(1) {
        None => 10,
        Some(val) => match val.parse::<isize>() {
            Ok(val) => val,
            Err(_) => {
                println!("argument is not a number");
                return;
            }
        },
    };

    println!("{}", run(arg));
}
